docs/src/explanations/chisel-enum.md

```diff
@@ -0,0 +1,129 @@
+
+## Importing the ChiselEnum module
+
+```scala mdoc:silent
+import chisel3._
+import chisel3.experimental.ChiselEnum
+```
+
+## Functionality and Examples
+
+Below we see ChiselEnum being used as mux select for a RISC-V core. While wrapping the object in a package is not required, it is highly recommended as it allows for the type to be used in multiple files more easily. 
+
+```scala mdoc:silent
+package CPUTypes {
+    object AluMux1Sel extends ChiselEnum {
+        val rs1out, pcout = Value
+        /**
+            "rs1out" -> 0.U,
+            "pcout"  -> 1.U
+        */
+    }
+}
+```
+
+Here we see a mux using the AluMux1Sel to select between different inputs. 
+
+```scala mdoc:silent
+import CPUTypes.AluMux1Sel
+import CPUTypes.AluMux1Sel._
+
+class AluMux1File(val dl_size: Int) extends Module {
+    val io = IO(new Bundle {
+        val aluMux1Sel =  Input( AluMux1Sel() )
+        val rs1Out     =  Input(Bits(dl_size.W))
+        val pcOut      =  Input(Bits(dl_size.W))
+        val aluMux1Out = Output(Bits(dl_size.W))
+    })
+
+    // Default value for aluMux1Out
+    io.aluMux1Out := 0.U
+
+    switch (io.aluMux1Sel) {
+        is (rs1out) {
+            io.aluMux1Out  := io.rs1Out
+        }
+        is (pcout) {
+            io.aluMux1Out  := io.pcOut
+        }
+    }
+}
+```
+```scala mdoc:verilog
+import 
+ChiselStage.emitVerilog.emitVerilog(new AluMux1File(dl_size = 32) )
+```
+
+
+ChiselEnum also allows for the user to define variables by passing in the value shown below. Note that the value must be increasing.
+
+```scala mdoc:silent
+object Opcode extends ChiselEnum {
+    val load  = Value(0x03.U) // i load  -> 000_0011
+    val imm   = Value(0x13.U) // i imm   -> 001_0011
+    val auipc = Value(0x17.U) // u auipc -> 001_0111
+    val store = Value(0x23.U) // s store -> 010_0011
+    val reg   = Value(0x33.U) // r reg   -> 011_0011
+    val lui   = Value(0x37.U) // u lui   -> 011_0111
+    val br    = Value(0x63.U) // b br    -> 110_0011
+    val jalr  = Value(0x67.U) // i jalr  -> 110_0111
+    val jal   = Value(0x6F.U) // j jal   -> 110_1111
+}
+```
+
+The user can 'jump' to a value and continue incrementing by passing a start point then using a regular Value assignment. 
+
+```scala mdoc:silent
+object BranchFunct3 extends ChiselEnum {
+    val beq, bne = Value
+    val blt = Value(4.U)
+    val bge, bltu, bgeu = Value
+    /**
+        "beq"  -> 0.U,
+        "bne"  -> 1.U,
+        "blt"  -> 4.U,
+        "bge"  -> 5.U,
+        "bltu" -> 6.U,
+        "bgeu" -> 7.U
+    */
+}
+```
+
+## Testing
+
+When testing your modules, the `.Type` and `.litValue` attributes allow for the the objects to be passed as parameters and for the value to be converted to BigInt type. Note that BigInts cannot be casted to Int with `.asInstanceOf[Int]`, they use their own methods like `toInt`. Please review the [scala.math.BigInt](https://www.scala-lang.org/api/2.12.5/scala/math/BigInt.html) page for more details!
```

Comment:
the example below is a nice example of `litValue`, can you show an example of `.Type` usage?

Also, does `sel.litValue == 0.U` also work? It seems more intuitive... actually it seems way more intuitive to actually use the Enum values in such a comparison. The below seems like a pattern that should be discouraged

Reply:
.Type is used in line 100. In the testbench, the .Type function is used to pass the ChiselEnum as a parameter since we do not want the literal value!

Reply:
`sel.litValue == 0.U` does not work since .litValue returns a BigInt and not a UInt. This block is pulled out of my testbench which tests that the ChiselEnum type is enumerating correctly and that all values match as expected.

Reply:
gotcha. I can't resolve this comment for some reason but you've answered my question